Source/Core/DolphinQt/NetPlay/NetPlayDialog.cpp

```diff
@@ -458,6 +462,57 @@ void NetPlayDialog::show(std::string nickname, bool use_traversal)
   UpdateGUI();
 }
 
+void NetPlayDialog::UpdateDiscordPresence()
+{
+#ifdef USE_DISCORD_PRESENCE
+  // both m_current_game and m_player_count need to be set for the status to be displayed correctly
+  if (m_player_count == 0 || m_current_game.empty())
+    return;
+
+  const auto use_default = [this]() {
+    Discord::UpdateDiscordPresence(m_player_count, Discord::SecretType::Empty, "", m_current_game);
+  };
+
+  if (g_TraversalClient)
+  {
+    const auto host_id = g_TraversalClient->GetHostID();
+    if (host_id == decltype(host_id)())
+      return use_default();
+
+    Discord::UpdateDiscordPresence(m_player_count, Discord::SecretType::RoomID,
+                                   std::string(host_id.begin(), host_id.end()), m_current_game);
+  }
+  else if (IsHosting())
+  {
+    if (m_exernal_ip_address.empty())
+    {
+      Common::HttpRequest request;
+      // ENet does not support IPv6, so IPv4 has to be used
+      request.UseIPv4();
+      Common::HttpRequest::Response response =
+          request.Get("https://ip.dolphin-emu.org/", {{"X-Is-Dolphin", "1"}});
+
+      if (!response.has_value())
+        return use_default();
+      m_exernal_ip_address = std::string(response->begin(), response->end());
+    }
+    const int port = Settings::Instance().GetNetPlayServer()->GetPort();
+
+    Discord::UpdateDiscordPresence(m_player_count, Discord::SecretType::IPAddress,
```

Comment:
Note: I am unfamiliar with Discord's internals.

I'm worried about the potential for users to leak their IP address by accident when they didn't intend to. Is the value of the secret ever shared with other Discord users before the host has approved the join request?

Reply:
I asked the Discord API server once
https://discordapp.com/channels/81384788765712384/381870553235193857/466009017630851075

I'll post a copy here at the end. Other then that, I think IP addresses are only sent to other users if the user accepts the other user, and Discord will then send the IP address to the other user. I'm also going to guess that sending the IP address to Discord is likly done using the WebSocket Secure protocol or HTTPS. However, I think it's best to ask a Discord dev like Mason ( @msciotti ) to make sure this info is correct.

I'll also now do some testing and searching to see if Dolphin ever saves the IP address to make sure it isn't easy to get without networking analyzing software.

Copy of Chat:
Sleepy Flower Girl
This place doesn't have a rich presence channel so I'll post here. For Discord rich presence joining feature, they want the secret to be hashed/encoded/encrypted. However, I need to use the hostname or ip address as the secret, any idea how to do this securely and properly?

Sharon
why do you have to use those as the secret

Sleepy Flower Girl
Discord sends me the secret during the join callback, and I need the ip address or hostname or domain name to connect to the right game server

Lachee
Sleepy
I have spoken to mason about this
You can use ip
Secret just needs to be what ever you need to connect

Sleepy Flower Girl
without needing to hash or encrypt?

Rensia 
the secrets are supposed to be hashes that you can then request from a server to retrieve proper details

Lachee
Encryption would be preferred, but not nessary as it is only sent once a JOIN has been approved
And rensia is correct, ideally it would have a master server

Reply:
We do have a master server, but that only applies in traversal mode. However, the concern here is only with direct connection leaking the host's IP before they've accepted the join request. As long as the Discord API is only sharing the secret once the request is accepted, it's fine, as IPs are exposed anyways once the connection is established in either mode, since this is a peer-to-peer model.